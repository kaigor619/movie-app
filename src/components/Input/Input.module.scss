.label {
  color: rgba(var(--text-primary-rgb), 0.6);
  font-size: 0.8rem;
  margin-bottom: 0.3rem;
  font-weight: 400;
}

.inputWrapper {
  position: relative;
  margin-bottom: 1.5rem;
  --bg-color: #2a353e;
  --bg-focus-color: #303d48;
  --border-color: #2a353e;
  --border-focus-color: #303d48;
  --text-color: rgba(var(--text-primary-rgb), 0.8);

  &[data-theme="light"] {
    --bg-color: white;
    --bg-focus-color: white;
    --border-color: #e2e2e2;
    --border-focus-color: black;
  }
}

.inputError {
  position: absolute;
  bottom: 100%;
  right: 0;
  color: var(--error-color);
  font-size: 0.7rem;
  transform: translateY(-0.2rem);
}

.input {
  padding: 0.5rem 1.2rem 0.5rem 0.5rem;
  background-color: var(--bg-color);
  border: none;
  border-radius: 0.3rem;
  color: var(--text-color);
  width: 100%;
  box-sizing: border-box;
  border: 0.1rem solid var(--border-color);
  font-size: 0.8rem;
  outline: none;
  transition: 0.2s background-color;

  &.error {
    border: 0.1rem solid var(--error-color);
  }

  &:focus {
    background-color: var(--bg-focus-color);
    border: 0.1rem solid var(--border-focus-color);
  }

  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }

  &[type="number"] {
    -moz-appearance: textfield;
  }
}

.microphone {
  color: rgba(var(--text-primary-rgb), 0.6);
  position: absolute;
  top: 50%;
  right: 0.5rem;
  transform: translateY(-50%);
  width: 0.9rem;
  height: auto;
  cursor: pointer;

  &.active {
    color: rgb(var(--text-primary-rgb));
    animation: microphone 0.3s linear 0s infinite alternate;
  }
}

@keyframes microphone {
  from {
    opacity: 1;
  }
  to {
    opacity: 0.5;
  }
}
