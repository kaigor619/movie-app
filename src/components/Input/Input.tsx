import { observer } from "mobx-react-lite";
import clsx from "clsx";
import { themeStore } from "stores/ThemeStore";

import styles from "./Input.module.scss";

interface Props {
  name: string;
  label?: string;
  type?: "number" | "text";
  errorMessage?: string;
  value: string;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onFocus?: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onBlur?: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

export const Input: React.FC<Props> = observer(
  ({ name, label, type = "text", errorMessage, onChange, onFocus, onBlur }) => {
    return (
      <>
        {label && (
          <label htmlFor={name} className={styles.label}>
            {label}
          </label>
        )}

        <div className={styles.inputWrapper} data-theme={themeStore.theme}>
          <input
            id={name}
            name={name}
            type={type}
            className={clsx(styles.input, {
              [styles.error]: !!errorMessage,
            })}
            onChange={onChange}
            onFocus={onFocus}
            onBlur={onBlur}
          />

          {errorMessage && (
            <span className={styles.inputError}>{errorMessage}</span>
          )}
        </div>
      </>
    );
  }
);
