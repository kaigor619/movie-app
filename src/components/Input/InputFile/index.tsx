export * from "./InputFile";
export * from "./InputFileController";
