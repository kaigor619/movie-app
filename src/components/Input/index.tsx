export * from "./Input";
export * from "./InputController";
export * from "./InputFile";
