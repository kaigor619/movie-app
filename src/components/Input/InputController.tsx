import { Controller, UseControllerProps } from "react-hook-form";

import { Input } from "./Input";

type InputProps = React.ComponentProps<typeof Input>;

type Props<T extends Record<string, any>> = Omit<
  InputProps,
  "value" | "onChange" | "onBlur" | "onFocus" | "errorMessage"
> &
  RequiredBy<UseControllerProps<T>, "control">;

export function InputController<T extends Record<string, any>>(
  props: Props<T>
): React.ReactElement<Props<T>> {
  return (
    <Controller
      name={props.name}
      control={props.control}
      rules={props.rules}
      render={({ field, fieldState }) => (
        <Input
          name={props.name}
          label={props.label}
          type={props.type}
          value={field.value}
          onChange={field.onChange}
          onBlur={field.onBlur}
          errorMessage={fieldState?.error?.message || ""}
        />
      )}
    />
  );
}
