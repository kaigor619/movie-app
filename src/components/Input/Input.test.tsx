import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Input } from "./Input";

describe("Input component", () => {
  const onChange = jest.fn();
  const onFocus = jest.fn();
  const mockInputValue = "Bruce";

  afterEach(() => {
    onChange.mockClear();
    onFocus.mockClear();
  });

  test("shows input value", () => {
    render(
      <Input
        name="name"
        value={mockInputValue}
        label="Name"
        errorMessage="Required"
        onChange={onChange}
        type="text"
      />
    );

    expect(screen.getByDisplayValue(mockInputValue)).toBeInTheDocument();
  });

  test("shows input label", () => {
    render(
      <Input
        name="name"
        value=""
        label="Name"
        errorMessage="Required"
        onChange={onChange}
        type="text"
      />
    );

    expect(screen.getByText(/Name/i)).toBeInTheDocument();
  });

  test("shows error message", () => {
    render(
      <Input
        name="name"
        value=""
        label="Name"
        errorMessage="Required"
        onChange={onChange}
        type="text"
      />
    );

    expect(screen.getByText(/Required/i)).toBeInTheDocument();
  });

  test("checks snapshot", () => {
    const { asFragment } = render(
      <Input
        name="name"
        value={mockInputValue}
        label="Name"
        errorMessage="Required"
        onChange={onChange}
        type="text"
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("checks onChange", async () => {
    render(<Input name="name" value="" onChange={onChange} />);

    const input = screen.getByRole("textbox");

    await userEvent.type(input, mockInputValue);

    expect(onChange).toHaveBeenCalledTimes(mockInputValue.length);
  });

  test("checks onFocus", async () => {
    render(
      <Input name="name" value="Ihor" onChange={onChange} onFocus={onFocus} />
    );

    const input = screen.getByRole("textbox");

    await userEvent.click(input);

    expect(onFocus).toHaveBeenCalledTimes(1);
  });
});
