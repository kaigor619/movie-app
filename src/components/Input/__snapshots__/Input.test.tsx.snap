// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component checks snapshot 1`] = `
<DocumentFragment>
  <label
    class="label"
    for="name"
  >
    Name
  </label>
  <div
    class="inputWrapper"
    data-theme="dark"
  >
    <input
      class="input error"
      id="name"
      name="name"
      type="text"
      value="Bruce"
    />
    <span
      class="inputError"
    >
      Required
    </span>
  </div>
</DocumentFragment>
`;
